// webdav/dist/node/response.js
import minimatch from "minimatch";
import { convertResponseHeaders } from "./tools/headers.js";
export function createErrorFromResponse(response, prefix = "") {
    const err = new Error(`${prefix}Invalid response: ${response.status} ${response.statusText}`);
    err.status = response.status;
    err.response = response;
    return err;
}
export function handleResponseCode(context, response) {
    const { status } = response;
    if (status === 401 && context.digest)
        return response;
    if (status >= 400) {
        const err = createErrorFromResponse(response);
        throw err;
    }
    return response;
}
export function processGlobFilter(files, glob) {
    return files.filter(file => minimatch(file.filename, glob, { matchBase: true }));
}
export function processResponsePayload(response, data, isDetailed = false) {
    return isDetailed
        ? {
            data,
            headers: response.headers ? convertResponseHeaders(response.headers) : {},
            status: response.status,
            statusText: response.statusText
        }
        : data;
}
